File: notificare/android/src/main/kotlin/re/notifica/flutter/events/NotificareEvent.kt
package re.notifica.flutter.events

import re.notifica.models.NotificareApplication
import re.notifica.models.NotificareDevice

internal sealed class NotificareEvent {

    abstract val type: Type
    abstract val payload: Any?

    enum class Type(val id: String) {
        READY(id = "ready"),
        UNLAUNCHED(id = "unlaunched"),
        DEVICE_REGISTERED(id = "device_registered"),
        URL_OPENED(id = "url_opened"),
    }

    class Ready(
        application: NotificareApplication
    ) : NotificareEvent() {
        override val type = Type.READY
        override val payload = application.toJson()
    }

    class Unlaunched : NotificareEvent() {
        override val type = Type.UNLAUNCHED
        override val payload: Nothing? = null
    }

    class DeviceRegistered(
        device: NotificareDevice
    ) : NotificareEvent() {
        override val type = Type.DEVICE_REGISTERED
        override val payload = device.toJson()
    }

    class UrlOpened(
        url: String,
    ) : NotificareEvent() {
        override val type = Type.URL_OPENED
        override val payload = url
    }
}

File: notificare/ios/Classes/Events/NotificareEvent.swift
//
//  NotificareEvent.swift
//  notificare
//
//  Created by Helder Pinhal on 03/12/2020.
//

import Foundation
import NotificareKit

enum NotificareEventType: String, CaseIterable {
    case ready
    case unlaunched
    case deviceRegistered = "device_registered"
    case urlOpened = "url_opened"
}

protocol NotificareEvent {
    var type: NotificareEventType { get }
    var payload: Any? { get }
}

class NotificareEventOnReady: NotificareEvent {
    let type: NotificareEventType
    let payload: Any?

    init(application: NotificareApplication) {
        self.type = .ready
        self.payload = try! application.toJson()
    }
}

class NotificareEventOnUnlaunched: NotificareEvent {
    let type: NotificareEventType
    let payload: Any?

    init() {
        self.type = .unlaunched
        self.payload = nil
    }
}

class NotificareEventOnDeviceRegistered: NotificareEvent {
    let type: NotificareEventType
    let payload: Any?

    init(device: NotificareDevice) {
        self.type = .deviceRegistered
        self.payload = try! device.toJson()
    }
}

class NotificareEventOnUrlOpened: NotificareEvent {
    let type: NotificareEventType
    let payload: Any?

    init(url: String) {
        self.type = .urlOpened
        self.payload = url
    }
}
